File: src/main/java/org/lotus/edu/math/Sequence.java
package org.lotus.edu.math;

import java.util.stream.IntStream;
import java.util.stream.Stream;

/**
 * Utility class to deal with some famous sequences.
 */
public class Sequence {

    private Sequence() {
    }

    /**
     * Returns stream of primes.
     * @return stream of primes
     */
    public static IntStream ofPrimes() {
        return IntStream.iterate(1, i -> i + 1).filter(Number::isPrime);
    }

    /**
     * Returns stream of Fibonacci numbers.
     * @return stream of Fibonacci numbers
     */
    public static IntStream ofFibonacciNumbers() {
        return Stream.iterate(new int[]{0, 1}, t -> new int[]{t[1], t[0] + t[1]}).mapToInt(t -> t[0]);
    }

    public static class Number {
        /**
         * Tests if provided {@code number} is prime.
         * @param number a number to test
         * @return {@code true} if the number is prime, {@code false} otherwise
         */
        public static boolean isPrime(int number) {
            if (number <= 1) {
                return false;
            }

            double testBoundary = Math.sqrt(number);
            return IntStream.rangeClosed(2, (int) testBoundary).noneMatch(i -> number % i == 0);
        }
    }
}

File: src/main/scala/org/lotus/edu/scala/math/Sequence.scala
package org.lotus.edu.scala.math

/**
 * Singleton object to deal with some famous sequences.
 */
object Sequence {

  /**
   * Returns stream of primes.
   * @return stream of primes
   */
  def ofPrimes(): Stream[Int] = {
    def sieve(s: Stream[Int]): Stream[Int] = s.head #:: sieve(s.tail filter (_ % s.head != 0))
    sieve(Stream.from(2))
  }

  /**
   * Returns stream of Fibonacci numbers.
   * @return stream of fibonacci numbers
   */
  def ofFibonacciNumbers(): Stream[Int] = {
    def appendTail(f1: Int, f2: Int): Stream[Int] = f1 #:: appendTail(f2, f1 + f2)
    appendTail(0, 1)
  }

  object Number {

    /**
     * Tests if provided number is prime,
     * @param number number to test
     * @return true if the number is prime, false otherwise
     */
    def isPrime(number: Int): Boolean = {
      if (number <= 1) false
      else (2 to math.sqrt(number).toInt) forall(i => number % i != 0)
    }
  }
}